### DISCUSSION TITLE
Type hint a decorator that returns a staticmethod

### DISCUSSION THREAD

**Discussion by zmievsa** [microsoft/pyright#5375] [Q&A]
I want to make the following decorator:

```python
def alter_response(model: type[BaseModel]) -> Callable[[Callable[[dict[str, Any]], None]], AlterResponseInstruction]:
    return staticmethod(AlterResponseInstruction)
```

So that I could write classes like this:

```python
class CollapseEventRequest:
    @alter_response(EventAPIResource)
    def alter_event_api_response(data: dict[str, Any]) -> None:
        data["request"] = data["request"]["id"]
```

Problem is: pyright marks the method as incorrect because it's first argument is of incorrect type (i.e. not Self) no matter how I type hint it.

I tried copying the type hints from typeshed's staticmethod and adding them to my AlterResponseInstruction, I tried type hinting that I return staticmethod, I even tried type hinting that I return Any but nothing helped :(

Is there anything I can do or pyright's team can do to allow for writing such decorators in pyright outside of "pyright: ignore" comments? Even "if TYPE_CHECKING" can't help here because my decorator accepts the "model" argument which staticmethod doesn't :(

**Comment by erictraut** [comment]
Could you provide a minimal self-contained sample? Your code above references symbols that are not defined.

I think the right answer is to not include `@static_method` in your decorator but instead apply your decorator to a static method. The `@staticmethod` decorator is necessarily special-cased in static type checkers. There's no way within the Python type system to differentiate between a `Callable` that represents a static method, a class method, or an instance method. This information needs to be determined by the static type checker, which means that it needs to special-case these decorators.

**Reply by zmievsa** [reply]
No worries, I closed the discussion because I have came to an "ok" solution:

I type hint alter_response to return a type[staticmethod[_P, None]] and then enforce the type of arguments of the decorated methods at runtime, thus getting the best of both worlds. It would be nice if we could avoid special-casing staticmethod or if we could have a "bound" argument in ParamSpec somehow but the solution above works for my use case.

Sorry for wasting your time.
